### DISCUSSION TITLE
__x.data return undefined

### DISCUSSION THREAD

**Discussion by Pulstarter** [alpinejs/alpine#1543] [7. Bugs (archived)] [answered]
I have a tabs component working fine on V2, but on V3 I get undefined when access to __x.$data on children nodes:

      function setTabs() {

            tabs = [...$refs.tabs.children];

            tabHeadings = tabs.map((tab, index) => {

                **tab.__x.$data.id = (index + 1); <---------- UNDEFINED ON V3**

                var name = tab.__x.$data.name;
                var index = tab.__x.$data.index;

                var errors = $(tab).find('.is-error').length;

                return {
                    name:name,
                    errors:errors,
                    index: index
                };

            });
        }

Any ideas on how to fix this?

**Comment by ryangjchandler** [comment, marked as answer]
@Pulstarter `__x.$data` is no longer a valid way of accessing the proxy in v3. You should be able to refactor your component to use nested scopes / `x-data` attributes.

**Reply by Pulstarter** [reply]
Thanks @ryangjchandler

So now, there is no way to access directly to a component x-data? I see on the docs that I can access from a children component to parent's x-data properties but not from parent to children like __x.$data.

**Reply by Pulstarter** [reply]
I rewrite the nested tab component with the new features and now works fine. Thanks!

**Reply by KevinBatdorf** [reply]
"How to access child component data" is a good question actually. I wonder if there's room for some sort of helper like `$children` or if this is an antipattern

**Reply by ryangjchandler** [reply]
I'd argue that its an antipattern now that shared / parent scopes are a thing.

**Reply by northern-mindset** [reply]
@ryangjchandler do you mean https://alpinejs.dev/essentials/state#global-state 
 and https://alpinejs.dev/essentials/state#nesting-data by “shared / parent scopes” respectively?

**Reply by ryangjchandler** [reply]
@anantakrishna That's correct.

**Reply by SebastianSchoeps** [reply]
> I rewrite the nested tab component with the new features and now works fine. Thanks!

Hey @Pulstarter , how exactly did you rewrite it? I have exactly the same issue...

**Reply by jelib3an** [reply]
I encountered the same issue - here's how I ended up refactoring it. To expose the child component data (in this case the name, in order to build the headings), I simply added it as a data attribute. Then in the child, I watch for whenever activeTab on the parent changes. 

Parent:
```html
<div x-data="{
    activeTab: 'Foo',
    tabHeadings: [],
}" x-init="() => {
    tabHeadings = [...$refs.tabs.children].map((tab) => tab.dataset.name);
}">
```
Child:
```html
<div x-data="{
    show: false,
    showIfActive() {
        this.show = ($el.dataset.name === this.activeTab);
    }
}" x-init="() => {
    showIfActive();
    $watch('activeTab', () => showIfActive());
}" x-show="show" data-name="Foo">
```

**Reply by ola-kiitan** [reply]
> I encountered the same issue - here's how I ended up refactoring it. To expose the child component data (in this case the name, in order to build the headings), I simply added it as a data attribute. Then in the child, I watch for whenever activeTab on the parent changes.
> 
> Parent:
> 
> ```
> <div x-data="{
>     activeTab: 'Foo',
>     tabHeadings: [],
> }" x-init="() => {
>     tabHeadings = [...$refs.tabs.children].map((tab) => tab.dataset.name);
> }">
> ```
> 
> Child:
> 
> ```
> <div x-data="{
>     show: false,
>     showIfActive() {
>         this.show = ($el.dataset.name === this.activeTab);
>     }
> }" x-init="() => {
>     showIfActive();
>     $watch('activeTab', () => showIfActive());
> }" x-show="show" data-name="Foo">
> ```

Could you please share your displayed html code of the tabs?
